### PR TITLE
keyboards/anavi: Add ANAVI Knobs 3

### PR DESCRIPTION
**Pull request by leon-anavi**
ANAVI Knobs 3 is a mini mechanical keyboard with 3 clickable rotary encoders, USB-C, RP2040 microcontroller and I2C slot for a display.

Signed-off-by: Leon Anavi <leon.anavi@konsulko.com>

## Description

Add a new small mechanical keyboard with RP2040. ANAVI Knobs 3 is a mini mechanical keyboard with 3 clickable rotary encoders, USB-C, RP2040 microcontroller and I2C slot for a display.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/anavi/knobs3/mcuconf.h [hunk 1]
```diff
@@ -0,0 +1,9 @@
+#pragma once
```
Comment (drashna) [inline keyboards/anavi/knobs3/mcuconf.h:1]: Could you add a GPL2+ compatible license header here?  All of the source files need to have a license explicitly listed.

For instance:
```c
// Copyright 2021 Your Name (@yourgithub)
// SPDX-License-Identifier: GPL-2.0-or-later
```

**Review by drashna — CHANGES_REQUESTED** [review]

File: keyboards/anavi/knobs3/mcuconf.h (hunk 1 above)
Comment (leon-anavi) [reply keyboards/anavi/knobs3/mcuconf.h:4]: Thank you for the code review. Sorry that I forgot about the license header. I have added it.

Thanks,
Leon

**Comment by leon-anavi** [comment]
Hi @drashna,

Following the successfully accepted GitHub pull request for Knob 1 I have rebased the patch for Knobs 3. These two keyboards are pretty similar and so is the code. Could you please have a look at the new version for Knobs 3?

Thanks, Leon

**Comment by leon-anavi** [comment]
Hi @drashna, @fauxpark,

Happy New Year! Best wishes for 2023. Sorry for bothering you but I would like to kindly remind about this GitHub pull request. I think all remarks from October have been resolved. Could you please let me know if any changes are required to get it merged?

Thanks,
Leon

**Review by drashna — APPROVED** [review]

**Comment by drashna** [comment]
Though, your files have a mix of spaces and tabs.  Could they be converted to all spaces?

**Comment by leon-anavi** [comment]
Hi @drashna,

Thank you for the code review and the feedback. I have converted the files to use only spaces.

Best regards,
Leon

File: keyboards/anavi/knobs3/config.h [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#pragma once
+
+#include "config_common.h"
+
+/* key matrix size */
+#define MATRIX_ROWS 1
+#define MATRIX_COLS 3
+
```
Comment (fauxpark) [inline keyboards/anavi/knobs3/config.h:11]:
```suggestion
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/anavi/knobs3/config.h (hunk 2 above)
Comment (leon-anavi) [reply keyboards/anavi/knobs3/config.h:11]: Hi @fauxpark,

If we remove these constants from `config.h`, we will need to use the real values to the keymaps in:

```
const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
    [_BASE] = LAYOUT_k3(
        KC_MUTE, RGB_TOG, RGB_MOD
    )
};
```
I don't insist but I prefer to keep the constants. However, it is your call.

Thanks,
Leon

Comment (fauxpark) [reply keyboards/anavi/knobs3/config.h:11]: The CLI can derive these from the structure of `DIRECT_PINS` (as you're using here), or the size of `MATRIX_ROW_PINS`/`MATRIX_COL_PINS`, and generates them automatically during compilation.

Comment (leon-anavi) [reply keyboards/anavi/knobs3/config.h:11]: This is really nice feature. I didn't know it. Thank you for letting me know. I will make the change and force push the updated commit.

File: keyboards/anavi/knobs3/keymaps/default/keymap.c [hunk 3]
```diff
@@ -0,0 +1,16 @@
+// Copyright 2022 Leon Anavi <leon@anavi.org>
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#include QMK_KEYBOARD_H
+
+//#include <stdio.h>
+
```
Comment (fauxpark) [inline keyboards/anavi/knobs3/keymaps/default/keymap.c:7]:
```suggestion
```

File: keyboards/anavi/knobs3/knobs3.c [hunk 4]
```diff
@@ -0,0 +1,73 @@
+// Copyright 2022 Leon Anavi <leon@anavi.org>
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#include "encoder.h"
```
Comment (fauxpark) [inline keyboards/anavi/knobs3/knobs3.c:4]:
```suggestion
#include "quantum.h"
```

File: keyboards/anavi/knobs3/info.json [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+            "resolution": 2
+        }
+        ]
+    },
+    "layouts": {
+        "LAYOUT_k3": {
+            "layout": [
+                { "label":"Mute", "x": 0, "y": 0 },
+                { "label":"RGB", "x": 0, "y": 1 },
+                { "label":"Animation", "x": 0, "y": 2 }
```
Comment (fauxpark) [inline keyboards/anavi/knobs3/info.json:63]:
```suggestion
                { "label":"Mute", "x": 0, "y": 0, "matrix": [0, 0] },
                { "label":"RGB", "x": 0, "y": 1, "matrix": [0, 1] },
                { "label":"Animation", "x": 0, "y": 2, "matrix": [0, 2] }
```

File: keyboards/anavi/knobs3/knobs3.h [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+#define ___ KC_NO
+
+// clang-format off
+ #define LAYOUT_k3(  \
+   K01, K02, K03 \
+ ) \
+ { \
+   { K01, K02, K03 } \
+ }
+// clang-format on
```
Comment (fauxpark) [inline keyboards/anavi/knobs3/knobs3.h:30]: This file can be deleted with the above changes to info.json.

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/anavi/knobs3/knobs3.h (hunk 6 above)
Comment (leon-anavi) [reply keyboards/anavi/knobs3/knobs3.h]: Thank you for the recommendations. I have applied the changes and forced pushed the amended git commit.

File: keyboards/anavi/knobs3/keymaps/default/keymap.c (hunk 3 above)
Comment (leon-anavi) [reply keyboards/anavi/knobs3/keymaps/default/keymap.c]: Thank you for the code review and sorry for the late feedback. I have applied the changed.

**Comment by leon-anavi** [comment]
Hi @drashna, @fauxpark,

Thank you for the code review and sorry for the late feedback on my side. I have just resolved and force pushed an updated version of the patch. Could you please have a look?

Thanks,
Leon

File: keyboards/anavi/knobs3/info.json [hunk 7]
```diff
@@ -0,0 +1,72 @@
+{
+    "keyboard_name": "Knobs 3",
+    "manufacturer": "ANAVI",
+    "url": "https://github.com/AnaviTechnology/anavi-knobs-3",
+    "maintainer": "leon-anavi",
+    "processor": "RP2040",
+    "bootloader": "rp2040",
+    "debounce": 5,
```
Comment (fauxpark) [inline keyboards/anavi/knobs3/info.json:8]:
```suggestion
    "matrix_pins": {
        "direct": [
            ["GP26", "GP29", "GP0"]
        ]
    },
```

File: keyboards/anavi/knobs3/config.h [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+#pragma once
+
+#include "config_common.h"
+
+/* Keyboard Matrix Assignments */
+// clang-format off
+#define DIRECT_PINS { \
+    { GP26, GP29, GP0 }  \
+}
+
```
Comment (fauxpark) [inline keyboards/anavi/knobs3/config.h:13]:
```suggestion
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

**Comment by leon-anavi** [comment]
Hi @fauxpark,

Thank you for the code review. I have applied the changed, amended and force pushed the commit.

Thanks, Leon

File: keyboards/anavi/knobs3/info.json [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+        },
+        {
+            "pin_a": "GP1",
+            "pin_b": "GP2",
+            "resolution": 2
+        }
+        ]
+    },
+    "layouts": {
+        "LAYOUT_k3": {
```
Comment (waffle87) [inline keyboards/anavi/knobs3/info.json:63]:
```suggestion
        "LAYOUT": {
```

File: keyboards/anavi/knobs3/keymaps/default/keymap.c [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#include QMK_KEYBOARD_H
+
+enum layer_names {
+    _BASE
+};
+
+const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
+    [_BASE] = LAYOUT_k3(
```
Comment (waffle87) [inline keyboards/anavi/knobs3/keymaps/default/keymap.c:11]:
```suggestion
const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
    [0] = LAYOUT(
```

File: keyboards/anavi/knobs3/readme.md [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+
+Mini mechanical keyboard with 3 clickable rotary encoders, USB-C, RP2040 microcontroller and I2C slot for mini OLED display.
+
+* Keyboard Maintainer: [Leon Anavi](https://github.com/leon-anavi)
+* Hardware Supported: ANAVI Knobs 3
+* Hardware Availability: [Crowd Supply](https://www.crowdsupply.com/anavi-technology/anavi-macro-pad-10), [GitHub repository](https://github.com/AnaviTechnology/anavi-knobs-3)
+
+Make example for this keyboard (after setting up your build environment):
+
+    qmk compile -kb anavi/knobs3 -km default
```
Comment (waffle87) [inline keyboards/anavi/knobs3/readme.md:11]:
```suggestion
    make anavi/knobs3:default

Flashing example for this keyboard:

    make anavi/knobs3:default:flash
```

**Review by waffle87 — CHANGES_REQUESTED** [review]

**Comment by leon-anavi** [comment]
Hi @waffle87,

Thank you for your code review. Following your suggestions I have applied the changes, amended and force pushed the commit.

Thanks, Leon

**Review by waffle87 — APPROVED** [review]

**Pull request merged**